File: sdks/csharp/PgChangeFeed.Client/Sse/PgChangeFeedSseClient.cs
using System.Net.Http.Headers;
using System.Runtime.CompilerServices;
using System.Text.Json;
using PgChangeFeed.Client.Http;
using PgChangeFeed.Client.Sse.Models;

namespace PgChangeFeed.Client.Sse;

/// <summary>
/// Client for the PG Change Feed live change stream over Server-Sent Events:
/// one method opens <c>GET /changes/stream</c>, assembles each SSE frame via
/// <see cref="SseFrameParser"/> and yields the ten message fields as
/// <see cref="Change"/> — the same form as the gRPC client
/// (<see cref="PgChangeFeed.Client.Grpc.PgChangeFeedGrpcClient.StreamChangesAsync"/>).
///
/// The <see cref="System.Net.Http.HttpClient"/> is passed in, not owned —
/// same as <see cref="PgChangeFeedHttpClient"/>, with one extra obligation
/// for this client: <see cref="System.Net.Http.HttpClient.Timeout"/> covers
/// the *entire* request/response lifetime for a streamed response, not just
/// the time to the response headers — a long-lived stream therefore needs
/// <c>Timeout = <see cref="Timeout.InfiniteTimeSpan"/></c> on the passed-in
/// client, otherwise the connection ends after the default timeout even
/// without a connection problem. The bearer token is sent in the
/// <c>Authorization</c> header as <c>Bearer &lt;token&gt;</c>, same as
/// <see cref="PgChangeFeedHttpClient"/>. There is no global or static state —
/// a process can hold several independently configured instances at once.
///
/// <b>Limits:</b> no delivery guarantee and no in-stream replay; a
/// disconnected or slow-reading consumer misses the affected messages
/// permanently. Missed changes remain recoverable through the read path
/// (<see cref="PgChangeFeedHttpClient.ReadChangesAsync"/>).
/// </summary>
public sealed class PgChangeFeedSseClient
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    private readonly HttpClient _httpClient;
    private readonly PgChangeFeedClientOptions _options;

    /// <summary>Creates the client over an <see cref="HttpClient"/> the caller owns.</summary>
    /// <param name="httpClient">The client used for the stream request; its timeout must be infinite for a long-lived stream.</param>
    /// <param name="options">The HTTP base URL and the bearer token.</param>
    public PgChangeFeedSseClient(HttpClient httpClient, PgChangeFeedClientOptions options)
    {
        ArgumentNullException.ThrowIfNull(httpClient);
        ArgumentNullException.ThrowIfNull(options);
        _httpClient = httpClient;
        _options = options;
    }

    /// <summary>
    /// Opens <c>GET /changes/stream</c> and yields every <see cref="Change"/>
    /// the server sends from connection time onward. A non-success response
    /// while opening the stream (e.g. a missing/unknown bearer token, or
    /// <c>503</c> when the stream is not available) becomes a typed
    /// <see cref="PgChangeFeedException"/>, the same set
    /// <see cref="PgChangeFeedHttpClient"/> throws. Once the stream is open,
    /// its end (regular server-side close, or an incomplete trailing frame)
    /// ends the enumeration without an exception — the same fire-and-forget
    /// behavior as
    /// <see cref="PgChangeFeed.Client.Grpc.PgChangeFeedGrpcClient.StreamChangesAsync"/>;
    /// a frame whose <c>data:</c> payload cannot be read still throws
    /// <see cref="PgChangeFeedMalformedResponseException"/>, because that is
    /// not a stream end.
    /// </summary>
    public async IAsyncEnumerable<Change> StreamChangesAsync(
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        var uri = new Uri(_options.Address, "/changes/stream");
        using var request = new HttpRequestMessage(HttpMethod.Get, uri);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _options.ApiToken);

        using var response = await _httpClient
            .SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken)
            .ConfigureAwait(false);

        if (!response.IsSuccessStatusCode)
        {
            var errorBody = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            throw BuildException((int)response.StatusCode, errorBody);
        }

        var statusCode = (int)response.StatusCode;
        using var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
        using var reader = new StreamReader(stream);

        while (true)
        {
            var frame = await SseFrameParser.ReadFrameAsync(reader, cancellationToken).ConfigureAwait(false);
            if (frame is null)
            {
                yield break;
            }
            yield return ParseChange(frame, statusCode);
        }
    }

    private static Change ParseChange(SseFrame frame, int statusCode)
    {
        Change? change;
        try
        {
            change = JsonSerializer.Deserialize<Change>(frame.Data, JsonOptions);
        }
        catch (JsonException ex)
        {
            throw new PgChangeFeedMalformedResponseException(
                statusCode,
                "PG Change Feed SSE stream delivered a frame whose data payload is not valid " +
                "JSON.",
                ex);
        }

        return change
            ?? throw new PgChangeFeedMalformedResponseException(
                statusCode,
                "PG Change Feed SSE stream delivered a frame whose data payload is empty or " +
                "null.");
    }

    private static PgChangeFeedException BuildException(int statusCode, string body)
    {
        var message = ExtractErrorMessage(body);
        return statusCode switch
        {
            400 => new PgChangeFeedBadRequestException(statusCode, message),
            401 => new PgChangeFeedUnauthorizedException(statusCode, message),
            403 => new PgChangeFeedForbiddenException(statusCode, message),
            404 => new PgChangeFeedNotFoundException(statusCode, message),
            500 => new PgChangeFeedServerErrorException(statusCode, message),
            _ => new PgChangeFeedUnexpectedStatusException(statusCode, message),
        };
    }

    private static string ExtractErrorMessage(string body)
    {
        try
        {
            var error = JsonSerializer.Deserialize<PgChangeFeed.Client.Http.Models.ErrorResponse>(body, JsonOptions);
            return error?.Error ?? body;
        }
        catch (JsonException)
        {
            return body;
        }
    }
}

File: sdks/kotlin/pgchangefeed-kotlin/src/main/kotlin/io/github/pt9912/pgchangefeed/sse/PgChangeFeedSseClient.kt
package io.github.pt9912.pgchangefeed.sse

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import io.github.pt9912.pgchangefeed.PgChangeFeedClientOptions
import io.github.pt9912.pgchangefeed.http.PgChangeFeedBadRequestException
import io.github.pt9912.pgchangefeed.http.PgChangeFeedException
import io.github.pt9912.pgchangefeed.http.PgChangeFeedForbiddenException
import io.github.pt9912.pgchangefeed.http.PgChangeFeedMalformedResponseException
import io.github.pt9912.pgchangefeed.http.PgChangeFeedNotFoundException
import io.github.pt9912.pgchangefeed.http.PgChangeFeedServerErrorException
import io.github.pt9912.pgchangefeed.http.PgChangeFeedUnauthorizedException
import io.github.pt9912.pgchangefeed.http.PgChangeFeedUnexpectedStatusException
import io.github.pt9912.pgchangefeed.http.TransportRequest
import io.github.pt9912.pgchangefeed.http.model.ErrorResponse
import io.github.pt9912.pgchangefeed.sse.model.Change
import java.net.http.HttpClient

/**
 * Client for the PG Change Feed live change stream over Server-Sent Events:
 * [streamChanges] opens `GET /changes/stream`, assembles each SSE frame via
 * [SseFrameParser] and yields the ten message fields as [Change]. Errors are
 * the typed [PgChangeFeedException] set that
 * [io.github.pt9912.pgchangefeed.http.PgChangeFeedHttpClient] throws (the
 * `{"error": "<text>"}` error body), not a second hierarchy.
 *
 * [streamChanges] returns a [Sequence]`<`[Change]`>`, not a
 * `kotlinx.coroutines.flow.Flow` like
 * [io.github.pt9912.pgchangefeed.grpc.PgChangeFeedGrpcClient.streamChanges]:
 * the generated gRPC coroutine stub offers a suspending, non-blocking call,
 * while `java.net.http.HttpClient.send()` (the JDK client
 * [io.github.pt9912.pgchangefeed.http.JdkHttpTransport] wraps) is an ordinary
 * blocking call, like every method of
 * [io.github.pt9912.pgchangefeed.http.PgChangeFeedHttpClient]. A [Sequence]
 * keeps the synchronous execution model of the HTTP client while staying
 * **cold**: nothing runs — no request is sent, no frame is parsed — until the
 * returned [Sequence] is iterated.
 *
 * The bearer token is sent in the `Authorization` header as
 * `Bearer <token>`, same as
 * [io.github.pt9912.pgchangefeed.http.PgChangeFeedHttpClient]. There is no
 * global or static state; a process can hold several independently configured
 * instances at once.
 *
 * **Limits:** no delivery guarantee and no in-stream replay — a disconnected
 * or slow-reading consumer misses the affected messages permanently. Missed
 * changes remain recoverable through the read path
 * (`io.github.pt9912.pgchangefeed.http.PgChangeFeedHttpClient.readChanges`).
 */
class PgChangeFeedSseClient internal constructor(
    private val transport: SseTransport,
    private val options: PgChangeFeedClientOptions,
) {
    /**
     * Public constructor — wraps the caller's `java.net.http.HttpClient` in the
     * real [JdkSseTransport]. The [HttpClient] is passed in, not owned — the
     * caller controls its lifetime, connection pooling and any redirect/proxy
     * configuration; this class never closes it. See [SseTransport] for why the
     * client depends on a transport interface rather than on [HttpClient]
     * directly.
     */
    constructor(httpClient: HttpClient, options: PgChangeFeedClientOptions) :
        this(JdkSseTransport(httpClient), options)

    /**
     * Opens `GET /changes/stream` and yields every [Change] the server sends
     * from connection time onward (fire-and-forget, no replay, one message per
     * row change in commit order). A non-success response while opening the
     * stream (e.g. a missing/unknown bearer token, or `503` when the stream is
     * not available) becomes a typed [PgChangeFeedException] once the returned
     * [Sequence] is iterated — this call itself never sends a request or
     * throws. Once the stream is open, its end (regular server-side close, or
     * an incomplete trailing frame) ends the sequence without an exception —
     * the same fire-and-forget behavior as
     * [io.github.pt9912.pgchangefeed.grpc.PgChangeFeedGrpcClient.streamChanges];
     * a frame whose `data:` payload cannot be read still throws
     * [PgChangeFeedMalformedResponseException], because that is not a stream
     * end.
     */
    fun streamChanges(): Sequence<Change> = sequence {
        val request = TransportRequest(
            method = "GET",
            url = options.address.toString().trimEnd('/') + "/changes/stream",
            headers = mapOf("Authorization" to "Bearer ${options.apiToken}"),
            body = null,
        )
        val response = transport.open(request)
        if (response.statusCode !in 200..299) {
            throw buildException(response.statusCode, drainToString(response.nextLine))
        }
        while (true) {
            val frame = SseFrameParser.readFrame(response.nextLine) ?: break
            yield(parseChange(frame.data, response.statusCode))
        }
    }

    private fun parseChange(data: String, statusCode: Int): Change {
        val change = try {
            gson.fromJson(data, Change::class.java)
        } catch (ex: JsonSyntaxException) {
            throw PgChangeFeedMalformedResponseException(
                statusCode,
                "PG Change Feed SSE stream delivered a frame whose data payload is not valid " +
                    "JSON.",
                ex,
            )
        }
        return change
            ?: throw PgChangeFeedMalformedResponseException(
                statusCode,
                "PG Change Feed SSE stream delivered a frame whose data payload is empty or " +
                    "null.",
            )
    }

    private fun buildException(statusCode: Int, body: String): PgChangeFeedException {
        val message = extractErrorMessage(body)
        return when (statusCode) {
            400 -> PgChangeFeedBadRequestException(statusCode, message)
            401 -> PgChangeFeedUnauthorizedException(statusCode, message)
            403 -> PgChangeFeedForbiddenException(statusCode, message)
            404 -> PgChangeFeedNotFoundException(statusCode, message)
            500 -> PgChangeFeedServerErrorException(statusCode, message)
            else -> PgChangeFeedUnexpectedStatusException(statusCode, message)
        }
    }

    private fun extractErrorMessage(body: String): String =
        try {
            gson.fromJson(body, ErrorResponse::class.java)?.error ?: body
        } catch (ex: JsonSyntaxException) {
            body
        }

    /**
     * Reassembles the remaining lines of a non-success response into a single
     * body string — the counterpart of
     * [io.github.pt9912.pgchangefeed.http.TransportResponse]'s already
     * buffered `body`. A non-success `GET /changes/stream` response carries a
     * single-line `{"error": "<text>"}` body, never a stream — joining every
     * remaining line with `\n` round-trips that single line unchanged.
     */
    private fun drainToString(nextLine: () -> String?): String {
        val builder = StringBuilder()
        while (true) {
            val line = nextLine() ?: break
            if (builder.isNotEmpty()) {
                builder.append('\n')
            }
            builder.append(line)
        }
        return builder.toString()
    }

    private companion object {
        val gson = Gson()
    }
}
